saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/serviceprovider/authentication/OpenSamlAuthenticationProvider.java

```diff
@@ -0,0 +1,338 @@
+		SignatureTrustEngine signatureTrustEngine = new ExplicitKeySignatureTrustEngine(
+				credentialsResolver,
+				DefaultSecurityConfigurationBootstrap.buildBasicInlineKeyInfoCredentialResolver()
+		);
+		SignaturePrevalidator signaturePrevalidator = new SAMLSignatureProfileValidator();
+		return new SAML20AssertionValidator(
+				conditions,
+				subjects,
+				statements,
+				signatureTrustEngine,
+				signaturePrevalidator
+		);
+	}
+
+	private Credential getVerificationCredential(X509Certificate certificate) {
+		return CredentialSupport.getSimpleCredential(certificate, null);
+	}
+
+	private Decrypter getDecrypter(Saml2X509Credential key) {
+		Credential credential = CredentialSupport.getSimpleCredential(key.getCertificate(), key.getPrivateKey());
+		KeyInfoCredentialResolver resolver = new StaticKeyInfoCredentialResolver(credential);
+		Decrypter decrypter = new Decrypter(null, resolver, this.saml.getEncryptedKeyResolver());
+		decrypter.setRootInNewDocument(true);
+		return decrypter;
+	}
+
+	private Assertion decrypt(Saml2AuthenticationToken idp, EncryptedAssertion assertion) {
+		Saml2Exception last = null;
+		for (Saml2X509Credential key : getDecryptionCredentials(idp)) {
+			final Decrypter decrypter = getDecrypter(key);
+			try {
+				return decrypter.decrypt(assertion);
+			}
+			catch (DecryptionException e) {
+				throw new Saml2Exception(e);
+			}
+		}
+		throw last;
+	}
+
+	private NameID decrypt(Saml2AuthenticationToken token, EncryptedID assertion) {
+		Saml2Exception last = null;
+		for (Saml2X509Credential key : getDecryptionCredentials(token)) {
+			final Decrypter decrypter = getDecrypter(key);
+			try {
+				return (NameID) decrypter.decrypt(assertion);
+			}
+			catch (DecryptionException e) {
+				last = new Saml2Exception(e);
+			}
+		}
+		throw last;
+	}
+
+	private List<Saml2X509Credential> getDecryptionCredentials(Saml2AuthenticationToken token) {
+		List<Saml2X509Credential> result = new LinkedList<>();
+		for (Saml2X509Credential c : token.getX509Credentials()) {
+			if (c.isDecryptionCredential()) {
+				result.add(c);
+			}
+		}
+		return result;
+	}
+
+	private List<X509Certificate> getVerificationKeys(Saml2AuthenticationToken token) {
+		List<X509Certificate> result = new LinkedList<>();
+		for (Saml2X509Credential c : token.getX509Credentials()) {
+			if (c.isSignatureVerficationCredential()) {
+				result.add(c.getCertificate());
+			}
+		}
+		return result;
+	}
+
+	public void setAuthoritiesMapper(GrantedAuthoritiesMapper authoritiesMapper) {
+		notNull(authoritiesMapper, "authoritiesMapper cannot be null");
+		this.authoritiesMapper = authoritiesMapper;
+	}
+
+	public void setResponseTimeValidationSkewMillis(Duration responseTimeValidationSkewMillis) {
+		this.responseTimeValidationSkewMillis = responseTimeValidationSkewMillis;
+	}
+
+	protected List<? extends GrantedAuthority> getAssertionAuthorities(Assertion assertion) {
```

Comment:
Make this private. First we prefer composition to inheritance and second we can add that feature later.